contracts/volt/VoltV2.sol

```diff
@@ -0,0 +1,440 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity =0.8.13;
+
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {VoltCoreRef} from "../refs/VoltCoreRef.sol";
+
+// Forked from Uniswap's UNI
+// Reference: https://etherscan.io/address/0x1f9840a85d5af5bf1d1762f925bdaddc4201f984#code
+
+contract VoltV2 is VoltCoreRef {
+    using SafeCast for *;
+
+    /// @notice EIP-20 token name for this token
+    // solhint-disable-next-line const-name-snakecase
+    string public constant name = "Volt";
+
+    /// @notice EIP-20 token symbol for this token
+    // solhint-disable-next-line const-name-snakecase
+    string public constant symbol = "VOLT";
+
+    /// @notice EIP-20 token decimals for this token
+    // solhint-disable-next-line const-name-snakecase
+    uint8 public constant decimals = 18;
+
+    /// @notice Total number of tokens in circulation
+    // solhint-disable-next-line const-name-snakecase
+    uint256 public totalSupply;
+
+    /// @notice Allowance amounts on behalf of others
+    mapping(address => mapping(address => uint256)) internal allowances;
+
+    /// @notice Official record of token balances for each account
+    mapping(address => uint256) internal balances;
+
+    /// @notice A record of each accounts delegate
+    mapping(address => address) public delegates;
+
+    /// @notice A checkpoint for marking number of votes from a given block
+    /// @dev  We use uint224 for the votes so we can pack this struct
+    /// into a single storage slot, therefore making transfers etc,
+    /// cheaper for our users, this follows the same pattern as
+    /// OpenZeppelins ERC20Votes which can be found here
+    /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Votes.sol
+    struct Checkpoint {
+        uint32 fromBlock;
+        uint224 votes;
+    }
+
+    /// @notice A record of votes checkpoints for each account, by index
+    mapping(address => mapping(uint32 => Checkpoint)) public checkpoints;
+
+    /// @notice The number of checkpoints for each account
+    mapping(address => uint32) public numCheckpoints;
```

Comment:
Could we combine balances and numCheckpoints into a single variable? This way, on each transfer, we can save an SSTORE if they have delegated to an address, or the address that's receiving the tokens has a delegate. then instead of 4 cold SSTORE operations per transfer with delegates, we could get down to 2 COLD SSTORE operations, and then do 2 additional WARM SSTORE operations, which would be considerably cheaper than a cold operation.

Reply:
Balances could be uint224 as that is the restriction on totalSupply, and numCheckpoints could remain as uint32, which is ~4 billion.

Reply:
We can try this design out, and see if it does anything in terms of gas efficiency.

Seems like this would only benefit us if transferring from and or to an address that has delegated, otherwise there is no benefit and there is a tiny addition to gas cost due to having to downcast the uint256's to the uint224's.